### PR TITLE
Make the checkpoint files be handled as Pegasus files.

### LINKED ISSUE #23: Use Pegasus's own checkpoint handling
A common pattern when building workflows is that you discover problems after a partial run which requires modification of the compute graph.  Here's an example of how this might work from this repository:

1. Run https://github.com/isi-vista/gaia-event-extraction/blob/master/gaia_event_extraction/pegasus/argument_attachment/argument_attachment_pipeline.py on https://github.com/isi-vista/gaia-event-extraction/blob/master/sample_params/argument_attachment/pegasus/bilstm_gold_mentions.params . This generates a `Workflow.dax`. This also populates what we will call our "experiment directory", whose path is given by parameters to the DAX generation script.  This experiment directory has a sub-directory for each "artifact" created in the course of the experiment execution, which typically corresponds to one job.  The are often (but don't have to be) arranged in a hierarchy which partly reflects the computation graph.  Each artifact sub-directory contains:
   - - a `___run.sh` which is what is actually submitted to the cluster. It sets up a virtual environment and executes the program.
   - a `__params.params` which contains the param file passed to the program 
2. `pegasus-plan -d Workflow.dax  --dir $HOME/workflows --relative-dir aa-full-run0001` . `--dir` should point to something in your home directory, rather than on `/nas/gaia`, because otherwise it crashes for some NFS-related reason.  This specifies the root directory under which what we will call "Pegasus run directories" live. We'll generate a new one of these each time we run Pegasus with a substantially different DAX.  The name of the Pegasus run directory for a given run is given by `--relative-dir`.
3. `pegasus-run  /nas/home/gabbard/workflows/aa-full-run0001` . This actually executes the workflow.  As each job finishes, a `__ckpt` file is written to the corresponding artifact folder in the experiment directory.  You can monitor the workflow with `pegasus-status -l /nas/home/gabbard/workflows/aa-full-run0001 -w 10`.  
4. Suppose the workflow crashes and you need to make a substantial change to how you build the DAX in order to fix it.  The problem I was running into which motivated the introduction of `__ckpt` files is that if you edit your DAX generation script in a way which causes the generated DAX to diverge too much, then an attempt to run `pegasus-plan` with the same Pegasus run directory caused a crash. However, if you ran it with a new run directory, it would repeat everything which had already been executed successfully on the previous run.  In our current approach, the DAX generation script will [detect if a `__ckpt` file is present for a job](https://github.com/isi-vista/vista-pegasus-wrapper/blob/master/pegasus_wrapper/workflow.py#L149). If so, it will not include it in the newly generated DAX and you will be able to continue with the steps below.
5. `pegasus-plan` again, but this time with a new run directory: ``pegasus-plan -d Workflow.dax  --dir $HOME/workflows --relative-dir aa-full-run0002` (note the new `relative-dir`). Then continue with `pegasus-status`, etc. 

The additional advantage of using `__ckpt` files is that it's not uncommon for us to have several workflows we wish to run independently which share sub-structures. For example, imagine I have one set of experiments which requires running BERT on all of Gigaword, which is very time consuming.  I might then have a second set of experiments which takes that as input and does something else with it. With this method, I can reuse the data from one to another even though they have totally independent DAXes.

There are some disadvantages: the user has to manage them by hand, which provides flexibility but also the chance for errors.  More importantly, it ties us to running from a shared filesystem, which limits our ability of using some of Pegasus's neat features like helping us automatically shift data around to run on e.g. AWS.

To ultimately drop the reliance on a shared filesystem, we need to be able to use Pegasus `File` objects for our inputs and outputs. However, Pegasus by default is set up expecting the input and output files to be specified as job arguments.  We would need to have a way of interpolating *the local path* to a file transferred by Pegasus on the execution machine into the parameter file ultimately passed to our executing script.  That's it's own issue.  For this issue, one of the Pegasus developers suggested that as a baby step we make the `___ckpt` files themselves Pegasus `File`s.  A job would depend on the `File`s corresponding the the checkpoint files of its "parent" jobs and its output would always be the `File` corresponding to its own `checkpoint`.  

I believe this would require modifying the `DependencyNode` class to wrap Pegasus files and also altering the generated `run.sh`s so that they write the checkpoint files *locally to the compute node* and then have them transferred back by Pegasus.

### LINKED ISSUE #29: Fix example workflow generation to submit to scavenge and not gaia partition

### PR DESCRIPTION
**Pull request by LichMaster98**
Closes #23 
Closes #29 

## Testing
1. Run `example_workflow_builder` script
2. `pegasus-plan` to create a final DAX
3. `pegasus-run` to generate output
4. Remove `sorted_nums.txt` and `working/jobs/sort/___ckpt`
5. Rerun steps 1 -> 3. Pegasus will not run the job to multiply nums because it already did and has record of it in the replica catalog.

### CODE REVIEW

File: pegasus_wrapper/artifact.py [hunk 1]
```diff
@@ -27,10 +27,16 @@ class DependencyNode:
     """
 
     job: Optional[Job] = attrib(validator=optional(instance_of(Job)), kw_only=True)
+    output_files: ImmutableSet[File] = attrib(
```
Comment (gabbard) [inline pegasus_wrapper/artifact.py:30]: Can you add a comment describing what this is for? (I know, but future readers might not)

File: pegasus_wrapper/pegasus_utils.py [hunk 2]
```diff
@@ -21,7 +21,11 @@
 
 
 def path_to_pegasus_file(
-    path: Path, *, site: str = "local", name: Optional[str] = None
+    path: Path,
+    *,
+    site: str = "local",
+    name: Optional[str] = None,
+    is_raw_input: bool = False
```
Comment (gabbard) [inline pegasus_wrapper/pegasus_utils.py:28]: What does this parameter do?

File: pegasus_wrapper/resource_request.py [hunk 3]
```diff
@@ -80,6 +80,9 @@ class SlurmResourceRequest(ResourceRequest):
     num_gpus: Optional[int] = attrib(
         validator=optional(in_(Range.at_least(0))), default=None, kw_only=True
     )
+    job_time: Optional[str] = attrib(
+        validator=optional(instance_of(str)), default="1:00:00", kw_only=True
```
Comment (gabbard) [inline pegasus_wrapper/resource_request.py:84]: It's generally best to avoid using strings for things which are not strings. :-). Perhaps make this `job_time_in_minutes: int` and then convert it to the string form in the Slurm-specific code?  Also, I think a higher default is better - say 24 hours.  Otherwise people who don't think to specify this will frequently have their jobs killed.

File: pegasus_wrapper/resource_request.py [hunk 4]
```diff
@@ -124,6 +128,10 @@ def apply_to_job(self, job: Job, *, log_file: Path, job_name: str) -> None:
                 self.memory if self.memory else _SLURM_DEFAULT_MEMORY
             ),
             stdout_log_path=log_file,
+            time=self.job_time,
+        )
+        logging.info(
```
Comment (gabbard) [inline pegasus_wrapper/resource_request.py:133]: This might be better done at the `debug` level to avoid too much clutter.

File: pegasus_wrapper/resources/sites.xml [hunk 5]
```diff
@@ -26,7 +26,7 @@
         <profile namespace="pegasus" key="style">glite</profile>
 
         <!--  batch queue to submit jobs to - both are required -->
-        <profile namespace="pegasus" key="queue">gaia</profile>
```
Comment (gabbard) [inline pegasus_wrapper/resources/sites.xml:29]: Why this change?

File: pegasus_wrapper/workflow.py [hunk 6]
```diff
@@ -47,16 +52,15 @@ class WorkflowBuilder:
     created_by: str = attrib(validator=instance_of(str), kw_only=True)
     _workflow_directory: Path = attrib(validator=instance_of(Path), kw_only=True)
     _namespace: str = attrib(validator=instance_of(str), kw_only=True)
+    _default_site: str = attrib(validator=instance_of(str), kw_only=True)
```
Comment (gabbard) [inline pegasus_wrapper/workflow.py:55]: Why was the order changed?

File: pegasus_wrapper/workflow.py [hunk 7]
```diff
@@ -67,6 +71,21 @@ class WorkflowBuilder:
     in the workflow graph. We compute this based on a job signature
     and only actually schedule the job once.
     """
+    _added_files: Set[File] = attrib(init=False, factory=set)
```
Comment (gabbard) [inline pegasus_wrapper/workflow.py:74]: Please add comments to this field and the following one explaining what they are for.

File: pegasus_wrapper/workflow.py [hunk 8]
```diff
@@ -67,6 +71,21 @@ class WorkflowBuilder:
...
+
+    _replica_catalog: Path = attrib(init=False)
+
+    def __attrs_post_init__(self) -> None:
+        # Do I need to check if it exists already? Or will users manually remove
```
Comment (gabbard) [inline pegasus_wrapper/workflow.py:79]: Prefer to do this work in the `from_parameters` factory method and just pass in the path to the replica catalog to the object constructor.

File: pegasus_wrapper/workflow.py [hunk 9]
```diff
@@ -67,6 +71,21 @@ class WorkflowBuilder:
...
+        # 'working' dir before rerunning?
+        replica_catalog_path = self._workflow_directory / "rc.dat"
+        if replica_catalog_path.exists():
+            replica_catalog_path.unlink()
+
+        replica_catalog_path.touch(mode=0o744)
+        object.__setattr__(self, "_replica_catalog", replica_catalog_path)
+
+        # self._job_graph.addFile(File(self._replica_catalog, ))
```
Comment (gabbard) [inline pegasus_wrapper/workflow.py:88]: Delete commented code.

File: pegasus_wrapper/workflow.py [hunk 10]
```diff
@@ -217,7 +233,28 @@ def run_python_on_parameters(
...
+        # See: https://github.com/isi-vista/vista-pegasus-wrapper/issues/25
+        checkpoint_pegasus_file = path_to_pegasus_file(
+            checkpoint_path, site=self._default_site, name=f"{ckpt_name}"
+        )
+
+        if checkpoint_pegasus_file not in self._added_files:
+            self._job_graph.addFile(checkpoint_pegasus_file)
+            self._added_files.add(checkpoint_pegasus_file)
+
+        if checkpoint_path.exists():
```
Comment (gabbard) [inline pegasus_wrapper/workflow.py:247]: This section is not clear to me.

File: scripts/multiply_by_x.py [hunk 11]
```diff
@@ -7,10 +10,14 @@ def main(params: Parameters):
     input_file_path = params.existing_file("input_file")
     output_file_path = params.creatable_file("output_file")
     x = params.integer("x")
+    logging.info("Reading from input file: %s", str(input_file_path.absolute()))
     with input_file_path.open() as input_file:
-        with output_file_path.open() as output_file:
+        with open(str(output_file_path.absolute()), "w") as output_file:
```
Comment (gabbard) [inline scripts/multiply_by_x.py:15]: You just want to pass "w" to `.open()` as it was previously written (it is the more idiomatic way of opening `pathlib` `Path`s)

File: scripts/sort_nums_in_file.py [hunk 12]
```diff
@@ -8,12 +10,16 @@
...
     input_file_path = params.existing_file("input_file")
     output_file_path = params.creatable_file("output_file")
+    logging.info("Reading from input file: %s", str(input_file_path.absolute()))
     with input_file_path.open() as input_file:
         nums = [int(x) for x in input_file]
 
     nums.sort()
 
-    output_file_path.write_text("\n".join(immutableset(nums)))
+    # time.sleep(60)
```
Comment (gabbard) [inline scripts/sort_nums_in_file.py:19]: Delete commented line

File: pegasus_wrapper/resources/sites.xml (hunk 5 above)
Comment (joecummings) [reply pegasus_wrapper/resources/sites.xml:29]: @LichMaster98 might have a better idea on this one

File: pegasus_wrapper/pegasus_utils.py (hunk 2 above)
Comment (joecummings) [reply pegasus_wrapper/pegasus_utils.py:28]: @LichMaster98

File: pegasus_wrapper/workflow.py (hunk 6 above)
Comment (joecummings) [reply pegasus_wrapper/workflow.py:55]: @LichMaster98

File: pegasus_wrapper/pegasus_utils.py (hunk 2 above)
Comment (LichMaster98) [reply pegasus_wrapper/pegasus_utils.py:28]: This parameter indicates that the raw is an input which is always going to exist on the file system at the specified location. If so, we can safely add the PFP to the pegasus DAX.

File: pegasus_wrapper/workflow.py (hunk 6 above)
Comment (LichMaster98) [reply pegasus_wrapper/workflow.py:55]: I think mypy was complaining about the order because it doesn't know that non-default params that are KW only can go after ones with defaults. This can be returned to the previous order.

File: pegasus_wrapper/pegasus_utils.py (hunk 2 above)
Comment (gabbard) [reply pegasus_wrapper/pegasus_utils.py:28]: What distinguishes a raw input from a non-raw input?

Comment (LichMaster98) [reply pegasus_wrapper/pegasus_utils.py:28]: A raw input is a file that doesn't come from the output of another job in the workflow.

Comment (gabbard) [reply pegasus_wrapper/pegasus_utils.py:28]: Thanks, @LichMaster98 . @joecummings , can you add this information to the docstring?

File: pegasus_wrapper/resources/sites.xml (hunk 5 above)
Comment (LichMaster98) [reply pegasus_wrapper/resources/sites.xml:29]: This change can be undone, this modification was made when investigating why jobs from the test workflow were always being submitted to the `gaia` partition rather than scavenge.

File: pegasus_wrapper/resource_request.py [hunk 13]
```diff
@@ -105,13 +116,38 @@ def unify(self, other: ResourceRequest) -> ResourceRequest:
             num_gpus=other.num_gpus if other.num_gpus is not None else self.num_gpus,
         )
 
+    def convert_time_to_slurm_format(self, job_time_in_minutes: int) -> str:
+        hours, mins = divmod(job_time_in_minutes, 60)
+        return f"{hours}:{str(mins)+'0' if mins < 10 else mins}:00"
+
+    def check_time_limits(self) -> None:
```
Comment (gabbard) [inline pegasus_wrapper/resource_request.py:123]: This ties the code too tightly to our current cluster configuration. I'd suggest instead documenting in the README what sort of errors the user will see if they specify too high a time limit.

File: pegasus_wrapper/resources/sites.xml (hunk 5 above)
Comment (gabbard) [reply pegasus_wrapper/resources/sites.xml:29]: If it works with this omitted entirely, I suggest deleting the line below rather than commenting it out.
